lib/ansible/utils/encrypt.py

```diff
@@ -33,31 +38,137 @@
 DEFAULT_PASSWORD_LENGTH = 20
 
 
-def do_encrypt(result, encrypt, salt_size=None, salt=None):
-    if PASSLIB_AVAILABLE:
+class BaseHash(object):
+    algo = namedtuple('algo', ['crypt_id', 'salt_size', 'implicit_rounds'])
+    algorithms = {
+        'md5_crypt': algo(crypt_id='1', salt_size=8, implicit_rounds=None),
+        'bcrypt': algo(crypt_id='2a', salt_size=22, implicit_rounds=None),
+        'sha256_crypt': algo(crypt_id='5', salt_size=16, implicit_rounds=5000),
+        'sha512_crypt': algo(crypt_id='6', salt_size=16, implicit_rounds=5000),
+    }
+
+    def __init__(self, algorithm, error_type):
+        self.algorithm = algorithm
+        self.error_type = error_type
+
+
+class CryptHash(BaseHash):
+    def __init__(self, algorithm, error_type):
+        super(CryptHash, self).__init__(algorithm, error_type)
+
+        if sys.platform.startswith('darwin'):
+            raise self.error_type("crypt.crypt not supported on Mac OS X/Darwin, install passlib python module")
+
+        if algorithm not in self.algorithms:
+            raise self.error_type("crypt.crypt does not support '%s' algorithm" % self.algorithm)
+        self.algo_data = self.algorithms[algorithm]
+
+    def hash(self, secret, **settings):
+        salt = self._salt(**settings)
+        rounds = self._rounds(**settings)
+        return self._encrypt(secret, salt, rounds)
+
+    def _salt(self, **settings):
+        salt = settings.get('salt')
+        salt_size = settings.get('salt_size')
+        if salt is None:
+            if salt_size is None:
+                salt_size = self.algo_data.salt_size
+            r = random.SystemRandom()
```

Comment:
Should not `ansible.utils.encrypt.random_password` be reused here ([example](https://github.com/ansible/ansible/blob/adc0efe10c5be833b5138e1d8b0d0316f87c5241/lib/ansible/plugins/lookup/password.py#L207)) ?

Reply:
I did some modifications:
* Added encrypt.random_salt which is called by password._random_salt and also used in encrypt.py
* random_salt does not do a to_text. I am not sure if that is actually needed.
  Before to_text was indirectly used via _gen_candiate_chars